### PR TITLE
Pagination: cleanup

### PR DESCRIPTION
**Pull request by nbramblett**
Addressed Tom's comments on previous CRs.

TEST=manual,auto

Eye test and unit tests.

### CODE REVIEW

File: src/ui/components/results/paginationcomponent.js [hunk 1]
```diff
@@ -164,7 +164,13 @@ export default class PaginationComponent extends Component {
   }
 
   updatePage (offset) {
-    this._onPaginate();
+    const results = this.core.globalStorage.getState(StorageKeys.VERTICAL_RESULTS) || {};
+    let oldOffset = this.core.globalStorage.getState(StorageKeys.SEARCH_OFFSET) || 0;
+    const limit = this.core.globalStorage.getState(StorageKeys.SEARCH_CONFIG).limit;
```
Comment (creotutar) [inline src/ui/components/results/paginationcomponent.js]: are we guaranteed to get a search config?

File: src/ui/components/results/paginationcomponent.js [hunk 2]
```diff
@@ -164,7 +164,13 @@ export default class PaginationComponent extends Component {
   }
 
   updatePage (offset) {
-    this._onPaginate();
+    const results = this.core.globalStorage.getState(StorageKeys.VERTICAL_RESULTS) || {};
+    let oldOffset = this.core.globalStorage.getState(StorageKeys.SEARCH_OFFSET) || 0;
```
Comment (creotutar) [inline src/ui/components/results/paginationcomponent.js]: very small nit: can this be a const?

File: src/ui/components/results/paginationcomponent.js [hunk 3]
```diff
@@ -164,7 +164,13 @@ export default class PaginationComponent extends Component {
...
+    const limit = this.core.globalStorage.getState(StorageKeys.SEARCH_CONFIG).limit;
+    const oldPageNumber = (oldOffset / limit) + 1;
```
Comment (creotutar) [inline src/ui/components/results/paginationcomponent.js]: can limit be 0?

File: src/ui/components/results/paginationcomponent.js (hunk 1 above)
Comment (nbramblett) [reply src/ui/components/results/paginationcomponent.js]: Limit defaults to 15 if not set by the user, iirc. I don't think it's possible to be null unless someone actively tried to null out limit, in which case they shouldn't really be using pagination.

File: src/ui/components/results/paginationcomponent.js (hunk 3 above)
Comment (nbramblett) [reply src/ui/components/results/paginationcomponent.js]: Hm, this is probably one of those cases where it could technically happen but would probably not ever happen (since it would mean you get 0 results). I'm somewhat of the opinion that if you set limit to 0 and then attempt to paginate then you deserve whatever implosion you get. They could also attempt to set limit to a negative number and bork themselves that way as well.

@tmeyer2115  do you have thoughts on this?

File: src/ui/components/results/paginationcomponent.js (hunk 1 above)
Comment (tmeyer2115) [reply src/ui/components/results/paginationcomponent.js]: We are guaranteed to get a SearchConfig. A default one is created if not supplied to Answers.init. The limit in this default SearchConfig will be 20.

File: src/ui/components/results/paginationcomponent.js (hunk 3 above)
Comment (tmeyer2115) [reply src/ui/components/results/paginationcomponent.js]: There is validation in the SDK that the limit supplied to a SearchConfig must be between 1 and 50.

File: src/ui/components/results/paginationcomponent.js [hunk 4]
```diff
@@ -164,7 +164,13 @@ export default class PaginationComponent extends Component {
   }
 
   updatePage (offset) {
-    this._onPaginate();
+    const results = this.core.globalStorage.getState(StorageKeys.VERTICAL_RESULTS) || {};
+    const oldOffset = this.core.globalStorage.getState(StorageKeys.SEARCH_OFFSET) || 0;
+    const limit = this.core.globalStorage.getState(StorageKeys.SEARCH_CONFIG).limit;
+    const oldPageNumber = (oldOffset / limit) + 1;
+    const pageNumber = (offset / limit) + 1;
+    const maxPageCount = Math.trunc((results.resultsCount - 1) / limit) + 1;
```
Comment (tmeyer2115) [inline src/ui/components/results/paginationcomponent.js]: We recompute the limit many times in this class. But, this quantity is effectively static (once the page has loaded). To avoid inefficiency, maybe we should compute limit once, in the constructor. I think we should also make a helper to compute maxPageCount. That way, we can refer to the helper and not have this logic written out explicitly in multiple places.

File: src/ui/components/results/paginationcomponent.js (hunk 3 above)
Comment (tmeyer2115) [reply src/ui/components/results/paginationcomponent.js]: Minor nit, but currentPageNumber?

File: src/ui/components/results/paginationcomponent.js [hunk 5]
```diff
@@ -239,18 +245,18 @@ export default class PaginationComponent extends Component {
         desktopBack: this._pinFirstAndLastPage && desktopBackLimit > 0,
```
Comment (tmeyer2115) [inline src/ui/components/results/paginationcomponent.js:259]: Do we need to check that it's greater than 1 here? If the backLimit is 1, wouldn't we not want ellipses?

File: tests/ui/components/results/paginationcomponent.js [hunk 6]
```diff
@@ -145,6 +145,6 @@ describe('rendering the page numbers', () => {
     const next = wrapper.find('.js-yxt-Pagination-next');
     expect(next).toHaveLength(1);
     next.first().simulate('click');
-    expect(paginate).toHaveBeenCalled();
```
Comment (tmeyer2115) [inline tests/ui/components/results/paginationcomponent.js:148]: Do we have test coverage for all the different edge cases that can occur in _allocate? That's one of the more complex methods in the Pagination component. We'll want to make sure we have coverage for the case in which _allocate will distribute pages evenly about the middle page and when it will be uneven since it's hit a boundary in some direction.

Comment (nbramblett) [reply tests/ui/components/results/paginationcomponent.js:148]: I've added some unit tests, it's a bit of a PITA to test comprehensively using enzyme, but I'll see about more thoroughly checking this.

File: src/ui/components/results/paginationcomponent.js (hunk 5 above)
Comment (nbramblett) [reply src/ui/components/results/paginationcomponent.js:259]: this is an interesting point, I suppose it would be a bit odd to see 1 ... 2 3 4 5 6 ... 10

I'll update this so that numbers have to actually be skipped for the ellipses to show up.

File: src/ui/components/results/paginationcomponent.js [hunk 7]
```diff
@@ -165,12 +172,11 @@ export default class PaginationComponent extends Component {
...
-    const oldOffset = this.core.globalStorage.getState(StorageKeys.SEARCH_OFFSET) || 0;
-    const limit = this.core.globalStorage.getState(StorageKeys.SEARCH_CONFIG).limit;
-    const oldPageNumber = (oldOffset / limit) + 1;
-    const pageNumber = (offset / limit) + 1;
-    const maxPageCount = Math.trunc((results.resultsCount - 1) / limit) + 1;
-    this._onPaginate(pageNumber, oldPageNumber, maxPageCount);
+    const currentOffset = this.core.globalStorage.getState(StorageKeys.SEARCH_OFFSET) || 0;
+    const currentPageNumber = (currentOffset / this._limit) + 1;
+    const newPageNumber = (offset / this._limit) + 1;
+    const maxPageCount = Math.trunc((results.resultsCount - 1) / this._limit) + 1;
```
Comment (tmeyer2115) [inline src/ui/components/results/paginationcomponent.js]: Use _computeMaxPage here

File: src/ui/components/results/paginationcomponent.js [hunk 8]
```diff
@@ -239,27 +253,32 @@ export default class PaginationComponent extends Component {
     }
 
     return {
-      ellipses: {
+      pinnedNumbers: {
```
Comment (tmeyer2115) [inline src/ui/components/results/paginationcomponent.js:256]: This is kind of confusing. What's the difference between pinnedNumbers and ellipses?

File: src/ui/components/results/paginationcomponent.js [hunk 9]
```diff
@@ -239,27 +253,32 @@ export default class PaginationComponent extends Component {
...
         mobileBack: this._pinFirstAndLastPage && mobileBackLimit > 0,
```
Comment (tmeyer2115) [inline src/ui/components/results/paginationcomponent.js:257]: If the backLimit is 1, will this show both Page '1' twice?

File: src/ui/templates/results/pagination.hbs [hunk 10]
```diff
@@ -38,11 +40,12 @@
     </a>
 
     <a
-      class="yxt-Pagination-link js-yxt-Pagination-link{{#unless ellipses.desktopBack}} desktop-hidden{{/unless}}{{#unless ellipses.mobileBack}} mobile-hidden{{/unless}}"
+      class="yxt-Pagination-link js-yxt-Pagination-link{{#unless pinnedNumbers.desktopBack}} desktop-hidden{{/unless}}{{#unless pinnedNumbers.mobileBack}} mobile-hidden{{/unless}}"
```
Comment (tmeyer2115) [inline src/ui/templates/results/pagination.hbs:43]: Would it be simpler to just conditionally render these tags instead of toggling hidden CSS classes on or off?

File: src/ui/components/results/paginationcomponent.js (hunk 8 above)
Comment (nbramblett) [reply src/ui/components/results/paginationcomponent.js:256]: ellipses will only show if there's a gap between the last page number and the pinned page number

File: src/ui/components/results/paginationcomponent.js (hunk 9 above)
Comment (nbramblett) [reply src/ui/components/results/paginationcomponent.js:257]: It will not.

File: src/ui/templates/results/pagination.hbs (hunk 10 above)
Comment (nbramblett) [reply src/ui/templates/results/pagination.hbs:43]: the limits are different values for mobile and desktop, and there's no way to conditionally render them from the handlebars while knowing the size of the browser (at least not without some extremely hacky js)

**Review by tmeyer2115 — APPROVED** [review]
